src/renderer/features/network/CustomRpcForm/model/add-custom-rpc-model.ts

```diff
@@ -0,0 +1,179 @@
+import { ExtendedChain } from '@entities/network';
+import { CustomRpcForm, NodeExistParam, RpcCheckResult } from '../lib/types';
+import { manageNetworkModel } from '@pages/Settings/Networks/model/manage-network-model';
+import { fieldRules } from '../lib/utils';
+
+const $addCustomRpcForm = createForm({
+  fields: {
+    name: {
+      init: '',
+      rules: fieldRules.name,
+    },
+    url: {
+      init: '',
+      rules: fieldRules.url,
+    },
+  },
+  validateOn: ['blur'],
+});
+
+const formInitiated = createEvent();
+const networkChanged = createEvent<ExtendedChain>();
+const rpcConnectivityChecked = createEvent<RpcCheckResult>();
+const nodeExistChecked = createEvent<boolean>();
+const processStarted = createEvent<boolean>();
+
+const $selectedNetwork = createStore<ExtendedChain | null>(null);
+const $rpcConnectivityResult = createStore<RpcCheckResult>(RpcCheckResult.INIT);
+const $isNodeExist = createStore<boolean>(false);
+const $isProcessStarted = createStore<boolean>(false);
+
+const checkRpcNodeFx = createEffect(
+  async ({ network, url }: { network: ExtendedChain | null; url: string }): Promise<RpcCheckResult | null> => {
+    if (!network) return null;
+
+    try {
+      rpcConnectivityChecked(RpcCheckResult.LOADING);
+      const validationResult = await networkService.validateRpcNode(network.chainId, url);
+
+      const RpcValidationMapping = {
+        [RpcValidation.INVALID]: RpcCheckResult.INVALID,
+        [RpcValidation.VALID]: RpcCheckResult.VALID,
+        [RpcValidation.WRONG_NETWORK]: RpcCheckResult.WRONG_NETWORK,
+      };
+
+      const result = RpcValidationMapping[validationResult];
+      rpcConnectivityChecked(result);
+
+      return result;
+    } catch (error) {
+      console.warn(error);
+
+      return null;
+    }
+  },
+);
+
+const saveRpcNodeFx = createEffect(
+  async ({
+    network,
+    form,
+  }: {
+    network: ExtendedChain | null;
+    form: CustomRpcForm;
+    rpcConnectivityResult: RpcCheckResult;
+    isNodeExist: boolean;
+  }) => {
+    if (!network) return;
+
+    manageNetworkModel.events.rpcNodeAdded({
+      chainId: network?.chainId,
+      rpcNode: {
+        name: form.name,
+        url: form.url,
+      },
+    });
+
+    processStarted(false);
+  },
+);
+
+const isNodeExistFx = createEffect(({ network, url }: NodeExistParam): boolean => {
+  if (!network) {
+    nodeExistChecked(false);
+
+    return false;
+  }
+
+  const defaultNodes = network.nodes;
+  const customNodes = network.connection.customNodes || [];
+
+  const result = defaultNodes.some(({ url: u }) => u === url) || customNodes.some(({ url: u }) => u === url);
+
+  nodeExistChecked(result);
+
+  return result;
+});
+
+const resetFormFx = createEffect(() => {
+  resetRpcValidationFx();
+  $addCustomRpcForm.reset;
+});
+
+const resetRpcValidationFx = createEffect(() => {
+  nodeExistChecked(false);
+  rpcConnectivityChecked(RpcCheckResult.INIT);
+});
```

Comment:
I must miss something, but to me, there's no simple way to call an effect with a sample. I went through the code a bit, and the `target` of a sample is always either an `effect`, or a `store`. The only place I saw `effect` being called was using a `spread` together with a `fn`. To me, this makes it hard to read an to maintain. But if you insist I should do this, I'll just do it :shrug: I just tend to prefer readability, also because as a newcomer to `effector` I find it extremely hard to review with those many samples.

Reply:
`clock` and `target` in sample are universal and you can use any from `store`, `effect`, or `event` for it

Reply:
In other words "any kind of Unit" and Unit is store/event/effect 😉

Reply:
But does something like this is ok? This looks wrong to me :)

```ts
sample({
  clock: saveRpcNodeFx.done,
  fn: () => false,
  target: processStarted,
});
```

Reply:
Answering my own question: It's probably not ok, the trick is to use `.reinit` on the resulting store, there's basically never a case in the code where a target is pointing at an event.